### PR TITLE
(vitineth/patch/api) Updating openapi specification

### PR DESCRIPTION
**Pull request by Vitineth**
This now handles almost all endpoints that I can think of but will likely need some more revisions

Updated readable spec is now available here:
* Stoplight (preferred): https://xiomi.stoplight.io/docs/uems-gateway-api/reference/kill-me.yaml
  * *ignore the url.. had some trouble making it*
* Swagger: https://app.swaggerhub.com/apis-docs/ents-crew/uems-gatway/0.0.1-alpha
  * its raising weird errors when saving so prefer the other one but I think it's up to date

### CODE REVIEW

**Comment by Lan2u** [comment]
In this what is an 'ent'?

**Comment by Vitineth** [comment]
> In this what is an 'ent'?

Short way to write "ents-state" so ents can have their own status on an event, distinct from a building status

**Comment by Lan2u** [comment]
> > In this what is an 'ent'?
> 
> Short way to write "ents-state" so ents can have their own status on an event, distinct from a building status

So as in its a flag set by the convener (or whoever) to indicate the stage that the event has been processed to?

**Comment by Vitineth** [comment]
> > > In this what is an 'ent'?
> > 
> > Short way to write "ents-state" so ents can have their own status on an event, distinct from a building status
> 
> So as in its a flag set by the convener (or whoever) to indicate the stage that the event has been processed to?

Yeah, something like "awaiting users", "waiting on requirements" or "awaiting techs"

**Comment by Lan2u** [comment]
> > > > In this what is an 'ent'?
> > > 
> > > 
> > > Short way to write "ents-state" so ents can have their own status on an event, distinct from a building status
> > 
> > 
> > So as in its a flag set by the convener (or whoever) to indicate the stage that the event has been processed to?
> 
> Yeah, something like "awaiting users", "waiting on requirements" or "awaiting techs"
> 
> Could do with a better name but wasn't sure what

I think implementing the system for adding/removing/getting these status names is a low priority as we can get very far simply by using a fixed list of allowed statuses when updating an event.

**Comment by Lan2u** [comment]
> > > > > In this what is an 'ent'?
> > > > 
> > > > 
> > > > Short way to write "ents-state" so ents can have their own status on an event, distinct from a building status
> > > 
> > > 
> > > So as in its a flag set by the convener (or whoever) to indicate the stage that the event has been processed to?
> > 
> > 
> > Yeah, something like "awaiting users", "waiting on requirements" or "awaiting techs"
> > Could do with a better name but wasn't sure what
> 
> I think implementing the system for adding/removing/getting these status names is a low priority as we can get very far simply by using a fixed list of allowed statuses when updating an event.

In line with this I think this API is a good starting point, it will be much easier to handle implementing if we take each part and split it into a separate issue to work on. I suspect there might be some stuff that is missed/doesn't work in practice so we will need to keep evaluating this as we go.

**Comment by Vitineth** [comment]
> > 
> > I think implementing the system for adding/removing/getting these status names is a low priority as we can get very far simply by using a fixed list of allowed statuses when updating an event.
> 
> In line with this I think this API is a good starting point, it will be much easier to handle implementing if we take each part and split it into a separate issue to work on. I suspect there might be some stuff that is missed/doesn't work in practice so we will need to keep evaluating this as we go.

Sounds good, basically get endpoints are the most important right now front-end wise and then can work up from there

**Comment by Lan2u** [comment]
> > > I think implementing the system for adding/removing/getting these status names is a low priority as we can get very far simply by using a fixed list of allowed statuses when updating an event.
> > 
> > 
> > In line with this I think this API is a good starting point, it will be much easier to handle implementing if we take each part and split it into a separate issue to work on. I suspect there might be some stuff that is missed/doesn't work in practice so we will need to keep evaluating this as we go.
> 
> Sounds good, basically get endpoints are the most important right now front-end wise and then can work up from there

Awesome - I'll approve this PR but the first thing we need to do is make some issues representing chunks of this

File: api.swagger.yaml [hunk 1]
```diff
@@ -249,3 +490,1399 @@ paths:
...
+              $ref: '#/components/schemas/CommentUpdate'
+        description: 'All properties are optional, any value provided will be updated'
+      tags:
+        - event
+        - comment
+    delete:
+      summary: Deletes the comment from the event
+      operationId: delete-events-id-comments-comment
+      responses:
+        '200':
```
Comment (Vitineth) [inline api.swagger.yaml:675]: According to https://docs.microsoft.com/en-us/azure/architecture/best-practices/api-design#delete-methods this should actually return `204`

**Comment by Vitineth** [comment]
Not to be addressed in this version but V2 of the API should using pagination (https://docs.microsoft.com/en-us/azure/architecture/best-practices/api-design#filter-and-paginate-data)

File: api.swagger.yaml (hunk 1 above)
Comment (Lan2u) [reply api.swagger.yaml:675]: Yep

**Comment by Lan2u** [comment]
> Not to be addressed in this version but V2 of the API should using pagination (https://docs.microsoft.com/en-us/azure/architecture/best-practices/api-design#filter-and-paginate-data)

Very much agree, I considered putting a limit on before but decided to leave it for now until we have a more firm understanding of how the system will actually behave

**Comment by Vitineth** [comment]
> > Not to be addressed in this version but V2 of the API should using pagination (https://docs.microsoft.com/en-us/azure/architecture/best-practices/api-design#filter-and-paginate-data)
> 
> Very much agree, I considered putting a limit on before but decided to leave it for now until we have a more firm understanding of how the system will actually behave

Fair enough, I figured we won't have enough data to make it worth it for a while so there is time to get a working platform then migrate to one with pagination

**Comment by Lan2u** [comment]
> > > Not to be addressed in this version but V2 of the API should using pagination (https://docs.microsoft.com/en-us/azure/architecture/best-practices/api-design#filter-and-paginate-data)
> > 
> > 
> > Very much agree, I considered putting a limit on before but decided to leave it for now until we have a more firm understanding of how the system will actually behave
> 
> Fair enough, I figured we won't have enough data to make it worth it for a while so there is time to get a working platform then migrate to one with pagination

Probably a good idea to put the parameters into the requests but leave them unused for now.

Also we might do well to include versioning in the API once we have a stable version.

**Comment by Vitineth** [comment]
I'm currently reading best practices so will have a couple comments as I find them. Should support each endpoint returning links to related endpoints as per the Microsoft recommendations. This can be added via the proposed link headers (https://tools.ietf.org/html/rfc8288) or through a more standard JSON response. This can be decided later because it's not a priority

**Comment by Vitineth** [comment]
> > > > Not to be addressed in this version but V2 of the API should using pagination (https://docs.microsoft.com/en-us/azure/architecture/best-practices/api-design#filter-and-paginate-data)
> > > 
> > > 
> > > Very much agree, I considered putting a limit on before but decided to leave it for now until we have a more firm understanding of how the system will actually behave
> > 
> > Fair enough, I figured we won't have enough data to make it worth it for a while so there is time to get a working platform then migrate to one with pagination
> 
> Probably a good idea to put the parameters into the requests but leave them unused for now.
> 
> Also we might do well to include versioning in the API once we have a stable version.

Good idea, the openapi spec currently lists the base API endpoint as `/api/v0` to allow for versioning in the future

File: api.swagger.yaml [hunk 2]
```diff
@@ -249,3 +490,1399 @@ paths:
...
+                          $ref: '#/components/schemas/CommentResponse'
+        '401':
+          $ref: '#/components/responses/UnauthorizedResponse'
+        '403':
+          $ref: '#/components/responses/ForbiddenResponse'
+        '404':
+          $ref: '#/components/responses/NotFoundIDResponse'
+      operationId: get-events-id-comments
+      description: Retrieves all comments related to the event with the specified ID
+    post:
```
Comment (Vitineth) [inline api.swagger.yaml:567]: Post requests should return the resource URL as a location header

**Comment by Lan2u** [comment]
With regards to the file handling (e.g. /events/{id}/files/{file}) what does {file} refer to? is it the raw file-name, a human-readable name or a hybrid?

**Comment by Lan2u** [comment]
Also, all API endpoints from now on should return 501 Not Implemented if they aren't implemented so we can distinguish between still intentional 404 and not-implemented 404.

**Comment by Vitineth** [comment]
> With regards to the file handling (e.g. /events/{id}/files/{file}) what does {file} refer to? is it the raw file-name, a human-readable name or a hybrid?
> 
> This would revolve around how we want to represent file objects to the front-end.
> 
> * Using the raw file names is potentially quite messy and hard to use (especially if we have a complex file storage hierarchy)
> * Using human readable names is much better but we will need to decide if those names must be always unique or only unique to the event. For example we might have a lx-plot.pdf for 2 different events and even though the name is the same the file may or may not be different.
> * We could also just use file ID's for everything as an abstraction over the actual file-path (and then map the ID -> the file at the back-end)

I was imagining a file ID but could also use a slug version of the filename (https://www.npmjs.com/package/slug) but considering this endpoint is only for metadata (except for uploading) it doesn't really matter

File: api.swagger.yaml (hunk 1 above)
Comment (Vitineth) [reply api.swagger.yaml]: Resolved via 800a3ba

File: api.swagger.yaml (hunk 2 above)
Comment (Vitineth) [reply api.swagger.yaml:560]: Resolved via 555d450

**Comment by Vitineth** [comment]
Re-requesting review for changes to DELETE and POST responses

File: api.swagger.yaml [hunk 3]
```diff
@@ -1791,6 +1765,12 @@ paths:
                     properties:
                       result:
                         $ref: '#/components/schemas/TopicResponse'
+          headers:
+            Location:
+              schema:
+                type: string
+                format: uri
+              description: The location at which the new resource can be found
```
Comment (Lan2u) [inline api.swagger.yaml:1773]: When you say location do you mean the URI of the API endpoint to GET the resource from?

Comment (Vitineth) [reply api.swagger.yaml:1773]: Yep so in this example it would have a `Location: /api/v0/topics/0` header or something like that

Comment (Lan2u) [reply api.swagger.yaml:1773]: Cool

**Pull request merged**